File: task16/task16.py
def sort(arr, start_index):
    size = len(arr)
    for curr in range(start_index, size-1):
        for next in range(curr, size):
            if arr[curr] > arr[next]:
                arr[curr], arr[next] = arr[next], arr[curr]

while True:
    num = list(input())
    size = len(num)

    is_found = False
    index = -1
    for i in range(size-1, 0, -1):
        for j in range(i-1, -1, -1):
            if num[j] < num[i]:
                num[j], num[i] = num[i], num[j]
                index = j
                is_found = True
                break
        if is_found:
            sort(num, index+1)
            break

    if is_found:
        res = ''.join(num)
        print(res)
    else:
        print("-1")
    print("")

File: task16/task16.cpp
#include <iostream>
#include <string>

using namespace std;

void sort(string* str, int startIndex) {
    int size = (*str).size();
    for (int current = startIndex; current < size - 1; current++) {
        for (int next = current; next < size; next++) {
            if ((*str)[current] > (*str)[next]) {
                swap((*str)[current], (*str)[next]);
            }
        }
    }
}

int main() {
    int num;
    cin >> num;
    string str = to_string(num);
    int size = str.size();
    
    
    bool isFound = false;
    int leftChangeIndex = -1, rightChangeIndex = -1;
    for (int i = size - 1; i > 0; i--) {
        for (int j = i - 1; j > leftChangeIndex; j--) {
            if (str[j] < str[i]) {
                if (leftChangeIndex < j) {
                    leftChangeIndex = j;
                    rightChangeIndex = i;
                }
                isFound = true;
                break;
            }
        }     
    }

    if (isFound) {
        swap(str[leftChangeIndex], str[rightChangeIndex]);
        sort(&str, leftChangeIndex + 1);
        cout << str << endl;
    }
    else {
        cout << -1 << endl;
    }
    
}